README.md
# Car Lease Calculator :red_car:

[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT)
[![Npm Version](https://img.shields.io/npm/v/lease-calculator?color=blue&logo=npm-version)](https://www.npmjs.com/package/lease-calculator)
![](https://github.com/ErezNagar/lease-calculator/workflows/Tests/badge.svg)

A simple auto lease calculator for calculating your monthly lease payment, APR, total lease cost, percentage off MSRP, etc.

See it used in the Auto Lease app: https://github.com/ErezNagar/lease-calculator-app

## Features

- Calculates car lease payments including total monthly payment, monthly payment pre-tax, total lease cost, APR, percentage off MSRP and drive-off fees.
- Since v4.0.0: Calculates car financing, including finance monthly payment, total finance cost and total finance interest
- Supports methods of taxation by state
- Includes acquisition fee for supported manufacturers
- Includes disposition fee for supported manufacturers
- Supports zero down payment
- Supports zero drive-off payments

## How To Use

```bash
npm install lease-calculator
```

```javascript
const leaseCalculator = new LeaseCalculator();
const lease = leaseCalculator.calculate({
  make: "Toyota",
  msrp: 23000,
  sellingPrice: 21000,
  rv: 13110,
  isRVPercent: false,
  mf: 0.00125,
  salesTax: 6.25,
  totalFees: 1200,
});

// Get the lease monthly payment
const monthlyPayment = lease.getMonthlyPayment();
```

## API

### `calculate(options: Object): LeaseResults`

Calculate lease payments. Returns LeaseResults object

Arguments:
`options` object, with the following attributes:

- `make: string = ""`
  Make of the vehicle, for calculating manufacturer-based fees.

- `msrp: number`
  Required, MSRP of the vehicle.

- `sellingPrice: number`
  Required, negotiated price of the vehicle.

- `rv: number`
  Required, residual value of the vehicle. If isRVPercent is true, value must be a percentage, e.g., if RV is 65%, rv should be 65.

- `isRVPercent: boolean = true`
  Whether the RV is an absolute value or a percentage of MSRP.

- `mf: number`
  Required, the money factor of the lease (e.g. 0.00125).

- `leaseTerm: number = 36`
  The length of the lease in months.

- `salesTax: number = 0`
  The state's sales tax in percentage.

- `totalFees: number = 0`
  Total fees of the lease.

- `rebates: number = 0`
  Total discount from dealer and manufacturer.

- `downPayment: number = 0`
  Down payment, if applicable.

- `taxMethod: TaxationMethod = TaxationMethod.TAX_ON_MONTHLY_PAYMENT`
  Method of taxation to apply, based on state.

- `isZeroDriveoff: boolean = false`
  Added in v1.3.0. Whether the lease should be calculated with zero drive-off amount. If true, all fees and taxes are rolled into the monthly payment.

#### `LeaseResults.getMonthlyPayment(): number`

Gets the monthly payment of the lease, including taxes.

#### `LeaseResults.getMonthlyPaymentPreTax(): number`

Gets the monthly payment of the lease, not including taxes.

#### `LeaseResults.getTotalLeaseCost(): number`

Gets the total cost of the lease. This includes all monthly payments, down payment, disposition fee, acquisition fee, dealer fees and lease taxes.

#### `LeaseResults.getDriveOffPayment(): number`

Gets total drive-off payment.

#### `LeaseResults.getDriveOffTax(): number`

Gets the tax amount on drive off payment.

#### `LeaseResults.getDiscountOffMsrpPercentage(): number | null`

Gets the discount off of MSRP, in percentage. Returns null if Selling Price >= MSRP (v3.0.0).

#### `LeaseResults.getMonthlyPaymentToMsrpPercentage(): number`

Gets the percentage of the monthly payment out of the MSRP.

#### `LeaseResults.getAPR()`

Gets the APR value of the lease.

#### `LeaseResults.getAcquisitionFee(): number`

Gets the acquisition fee value by brand. If no brand sepcified, returns 0.

#### `LeaseResults.getDispositionFee(): number`

Gets the disposition fee value by brand. If no brand sepcified, returns 0.

#### `LeaseResults.getRVValue(): number`

Gets the residual value of the lease.

#### `LeaseResults.getRVPercentage(): number`

Gets the residual value of the lease in percentage.

#### `LeaseResults.getDepreciation(): number`

Gets the total depreciation value of the lease. Added in v2.1.0.

#### `LeaseResults.getBaseMonthlyPayment(): number`

Gets the base monthly payment. Added in v2.1.0.

#### `LeaseResults.getRentCharge(): number`

Gets the rent charge value. Added in v2.1.0.

#### `LeaseResults.getTotalInterest(): number`

Gets total interest for the lease. Added in v2.1.0.

#### `LeaseResults.getMonthlyTax(): number`

Gets the monthly tax value. Applicable on for taxatino method is TAX_ON_MONTHLY_PAYMENT. Otherwise, returns 0. Added in v2.1.0.

#### `LeaseResults.getTotalTax(): number`

Gets the total tax for the lease. Added in v2.1.0.

#### `LeaseResults.getDriveOffPaymentBreakdown(): object[] | null`

Returns a list of all drive-off payments. Returns null if isZeroDriveoff is true.

#### `LeaseResults.getFinanceMonthlyPayment(): number`

Returns the finance monthly payment, inc. tax.

#### `LeaseResults.getFinanceTotalCost(): number`

Returns the total cost of finance. Comprised of the monthly payment over the life of the load plus down payment and any trade-in value.

#### `LeaseResults.getTotalAmountFinanced(): number`

Returns the total financed amount. Comprised of the selling price plus any fees and taxes minus down payment, trade-in value and any rebates.

#### `LeaseResults.getFinanceTotalInterest(): number`

Returns the total interest paid for finance.

### `calculateFinance (options: Object): FinanceResults`

Since v4.0.0. Calculates finance payments. Returns FinanceResults object.

Arguments:
`options` object, with the following attributes:

- `financeTerm: number`
  Required, the length of the loan, in months.

- `sellingPrice: number`
  Required, negotiated price of the vehicle.

- `APR: number`
  Required, the Annual Percentage Rate of the loan, in percentage (e.g. 4.5%).

- `salesTax :number = 0`,
  The state's sales tax, in percentage.

- `taxableFees :number = 0`,
  Total taxable fees (non-governmental fees such as dealer doc fees and any dealer add-ons)

- `untaxableFees :number = 0`,
  Total fees that are not taxable (government-related fees such as license and registration.

- `downPayment :number = 0`,
  Down payment, if applicable.

- `tradeIn :number = 0`,
  Trade-in value, if applicable.

- `rebates :number = 0`,
  Total discount from manufacturer.

#### `FinanceResults.getFinanceMonthlyPayment(): number`

Gets the financing monthly payment, including tax.

#### `FinanceResults.getFinanceTotalCost(): number`

Gets the total cost of the loan. Comprised of the monthly payment over the life of the loan plus down payment and any trade-in value.

#### `FinanceResults.getTotalAmountFinanced(): number`

Gets the total financed amount. Comprised of the selling price plus any fees and taxes.

#### `FinanceResults.getFinanceTotalInterest(): number`

Gets the total interest paid for duration of the loan.

## Supported manufacturers

Lease Calculator supports acquisition and disposition fee calculation for the following brands:

Acura, Alfa Romeo, Audi, Bmw, Buick, Cadillac, Chevrolet, Chrysler, Dodge, Fiat, Ford, Genesis, GMC, Honda, Hyundai, Infiniti, Jaguar, Jeep, Kia, land rover, Lexus, Lincoln, Mini, Mazda, Mercedes benz, Mitsubishi, Nissan, Ram, Scion, Smart, Subaru, Tesla, Toyota, Volkswagen, Volvo


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/).

## [4.1.0](https://github.com/ErezNagar/lease-calculator/compare/v4.0.0...v4.1.0) - 4/9/2023

### Updated

- Updated acquisition and disposition fee values

## [4.0.0](https://github.com/ErezNagar/lease-calculator/compare/v3.0.0...v4.0.0) - 17/9/2022

### Added

- `calculateFinance (options: Object): FinanceResults`: added new functionality to calculate finance payments. Returns a `FinanceResults` object:

  `FinanceResults.getFinanceMonthlyPayment(): number`: Gets the financing monthly payment, including tax.

  `FinanceResults.getFinanceTotalCost(): number`: Gets the total cost of the loan. Comprised of the monthly payment over the life of the loan plus down payment and any trade-in value.

  `FinanceResults.getTotalAmountFinanced(): number`: Gets the total financed amount. Comprised of the selling price plus any fees and taxes.

  `FinanceResults.getFinanceTotalInterest(): number`: Gets the total interest paid for duration of the loan.

- `LeaseResults.getDriveOffTax()` when calculating a lease - Gets the tax amount on drive off payment.

### Updated

- `LeaseResults.getDriveOffPaymentDetails()` updated to `LeaseResults.getDriveOffPaymentBreakdown()`

## [3.0.0](https://github.com/ErezNagar/lease-calculator/compare/v2.2.0...v3.0.0) - 08/7/2021

### Fixed

- Fixed getDiscountOffMsrpPercentage() to return null when Selling Price
  - `getDriveOffPaymentBreakdown(): object[] | null`

## [2.2.0](https://github.com/ErezNagar/lease-calculator/compare/v2.1.0...v2.2.0) - 04/6/2021

### Added

- An ability to get a detailed list of all drive-off payments
  - `getDriveOffPaymentBreakdown(): object[] | null`

## [2.1.0](https://github.com/ErezNagar/lease-calculator/compare/v2.0.0...v2.1.0) - 31/1/2021

### Added

- More granular details on different price components:
  - `getDepreciation(): number` - Gets the total depreciation value of the lease
  - `getBaseMonthlyPayment(): number` - Gets the base monthly payment
  - `getRentCharge(): number` - Gets the rent charge value
  - `getTotalInterest(): number` - Gets total interest for the lease
  - `getMonthlyTax(): number` - Gets the monthly tax value
  - `getTotalTax(): number` - Gets the total tax for the lease

## [2.0.0](https://github.com/ErezNagar/lease-calculator/compare/v1.2.2...v2.0.0) - 3/1/2021

### Added

- Support for zero drive-off

### Updated

- Acquisition fee is no longer capitalized by default
- Acquisition fee and disposition fee are included in drive-off payment when taxMethod = TaxationMethod.TAX_ON_TOTAL_LEASE_PAYMENT
- Total fees and rebates calculation

### Removed

- Support for trade-in value

## [1.2.1](https://github.com/ErezNagar/lease-calculator/compare/v1.2.0...v1.2.1) - 8/11/2020

### Added

- TypeScript support

## [1.2.0](https://github.com/ErezNagar/lease-calculator/compare/v1.1.0...v1.2.0) - 27/7/2020

### Added

- Support for different taxation methods:
  Tax levied on monthly payment (most states)
  Tax levied on selling/sales price (Virginia)
  Tax levied on total lease ppayment (New York, New Jersy, Minnesota, Ohio, Georgia)

## [1.1.0](https://github.com/ErezNagar/lease-calculator/compare/v1.0.1...v1.1.0) - 29/6/2020

### Added

- Acquisition and disposition fee support for 35 manufacturers:
  calculate() now accepts a `make` field for calculating manufacturer-based fees.
- getAcquisitionFeeValue(), returns acquisition fee value
- getDispositionFeeValue(), returns disposition fee value

## [1.0.1](https://github.com/ErezNagar/lease-calculator/compare/v1.0.0...v1.0.1) - 27/4/2020

### Fixed

- Removed required support for the experimental classProperties syntax
- Fixed package.json/main property

## [1.0.0](https://github.com/ErezNagar/lease-calculator/compare/v0.1.0...v1.0.0) - 27/4/2020

- Version official version

## [0.1.0](https://github.com/ErezNagar/lease-calculator/releases/tag/v0.1.0) - 14/4/2020

Initial version


tests/constants.js
import { TaxationMethod } from "../src/constants";

export const DUMMY_LEASE_ZERO_DOWN_WITH_TAX_ON_MONTHLY_PAYMENT = {
  make: "Toyota",
  msrp: 23000,
  sellingPrice: 21000,
  rv: 13110,
  isRVPercent: false,
  mf: 0.00125,
  leaseTerm: 36,
  salesTax: 10.25,
  totalFees: 1200,
  rebates: 500,
};

export const DUMMY_LEASE_WITH_DOWN_WITH_TAX_ON_MONTHLY_PAYMENT = {
  ...DUMMY_LEASE_ZERO_DOWN_WITH_TAX_ON_MONTHLY_PAYMENT,
  rebates: 0,
  downPayment: 1700,
};

export const DUMMY_LEASE_WITH_PERCENTAGE_RV = {
  ...DUMMY_LEASE_ZERO_DOWN_WITH_TAX_ON_MONTHLY_PAYMENT,
  rv: 57,
  isRVPercent: true,
};

export const RV_VALUE = 13110;
export const RV_PERCENTAGE = 57;
export const ACQUISITION_FEE_TOYOTA = 650;
export const DISPOSITION_FEE_TOYOTA = 350;

export const WHEN_TAXED_ON_MONTHLY_PAYMENT = {
  APR: 3,
  OFF_MSRP: 8.7,
  MSRP_PERCENTAGE: 1.19,
  BASE_PAYMENT: 205.28,
  RENT_CHARGE: 42.01,
  PAYMENT_ZERO_DOWN_PRE_TAX: 247.29,
  PAYMENT_ZERO_DOWN_ZERO_DRIVEOFF_PRE_TAX: 310.8,
  PAYMENT_ZERO_DOWN: 272.64,
  PAYMENT_WITH_DOWN: 234.23,
  PAYMENT_ZERO_DOWN_ZERO_DRIVEOFF: 342.65,
  TOTAL_LEASE_COST_ZERO_DOWN: 12255.82,
  TOTAL_LEASE_COST_WITH_DOWN: 12696.29,
  TOTAL_LEASE_COST_ZERO_DOWN_ZERO_DRIVEOFF: 12342.83,
  DRIVEOFF_AMOUNT: 2363.51,
};

export const DUMMY_LEASE_ZERO_DOWN_WITH_TAX_ON_SALES_PRICE = {
  ...DUMMY_LEASE_ZERO_DOWN_WITH_TAX_ON_MONTHLY_PAYMENT,
  taxMethod: TaxationMethod.TAX_ON_SALES_PRICE,
};

export const DUMMY_LEASE_WITH_DOWN_WITH_TAX_ON_SALES_PRICE = {
  ...DUMMY_LEASE_WITH_DOWN_WITH_TAX_ON_MONTHLY_PAYMENT,
  taxMethod: TaxationMethod.TAX_ON_SALES_PRICE,
};

export const WHEN_TAXED_ON_SALES_PRICE = {
  MSRP_PERCENTAGE: 1.08,
  PAYMENT_ZERO_DOWN_PRE_TAX: 282.12,
  PAYMENT_ZERO_DOWN_ZERO_DRIVEOFF_PRE_TAX: 372.52,
  PAYMENT_ZERO_DOWN: 247.29,
  PAYMENT_WITH_DOWN: 212.46,
  PAYMENT_ZERO_DOWN_ZERO_DRIVEOFF: 372.52,
  TOTAL_LEASE_COST_ZERO_DOWN: 13254.95,
  TOTAL_LEASE_COST_WITH_DOWN: 13700.95,
  TOTAL_LEASE_COST_ZERO_DOWN_ZERO_DRIVEOFF: 13388.05,
  DRIVEOFF_AMOUNT: 4249.79,
};

export const DUMMY_LEASE_ZERO_DOWN_WITH_TAX_ON_LEASE_PAYMENT = {
  ...DUMMY_LEASE_ZERO_DOWN_WITH_TAX_ON_MONTHLY_PAYMENT,
  taxMethod: TaxationMethod.TAX_ON_TOTAL_LEASE_PAYMENT,
};

export const DUMMY_LEASE_WITH_DOWN_WITH_TAX_ON_LEASE_PAYMENT = {
  ...DUMMY_LEASE_WITH_DOWN_WITH_TAX_ON_MONTHLY_PAYMENT,
  taxMethod: TaxationMethod.TAX_ON_TOTAL_LEASE_PAYMENT,
};

export const WHEN_TAXED_ON_LEASE_PAYMENT = {
  MSRP_PERCENTAGE: 1.08,
  PAYMENT_ZERO_DOWN_PRE_TAX: 300.99,
  PAYMENT_ZERO_DOWN_ZERO_DRIVEOFF_PRE_TAX: 348.98,
  PAYMENT_ZERO_DOWN: 247.29,
  PAYMENT_WITH_DOWN: 212.46,
  PAYMENT_ZERO_DOWN_ZERO_DRIVEOFF: 348.98,
  TOTAL_LEASE_COST_ZERO_DOWN: 12291.7,
  TOTAL_LEASE_COST_WITH_DOWN: 12732.16,
  TOTAL_LEASE_COST_ZERO_DOWN_ZERO_DRIVEOFF: 12564.24,
  DRIVEOFF_AMOUNT: 3286.54,
};

// Finance constants

export const DUMMY_FINANCE_WITH_ZERO_DOWN_WITH_ZERO_TAX_WITH_ZERO_FEES = {
  sellingPrice: 49000,
  APR: 1.5,
  financeTerm: 60,
};

export const DUMMY_FINANCE_WITH_DOWN_PAYMENT_WITH_ZERO_TAX_WITH_ZERO_FEES = {
  ...DUMMY_FINANCE_WITH_ZERO_DOWN_WITH_ZERO_TAX_WITH_ZERO_FEES,
  downPayment: 2000,
};

export const DUMMY_FINANCE_WITH_ZERO_DOWN_WITH_TAX_WITH_ZERO_FEES = {
  ...DUMMY_FINANCE_WITH_ZERO_DOWN_WITH_ZERO_TAX_WITH_ZERO_FEES,
  salesTax: 6.25,
};

export const DUMMY_FINANCE_WITH_ZERO_DOWN_WITH_ZERO_TAX_WITH_FEES = {
  ...DUMMY_FINANCE_WITH_ZERO_DOWN_WITH_ZERO_TAX_WITH_ZERO_FEES,
  untaxableFees: 500,
  taxableFees: 1000,
};

export const DUMMY_FINANCE_WITH_DOWN_PAYMENT_WITH_TAX_WITH_FEES = {
  ...DUMMY_FINANCE_WITH_ZERO_DOWN_WITH_ZERO_TAX_WITH_ZERO_FEES,
  downPayment: 2000,
  salesTax: 6.25,
  untaxableFees: 500,
  taxableFees: 1000,
};

export const DUMMY_FINANCE_WITH_TRADE_IN = {
  ...DUMMY_FINANCE_WITH_ZERO_DOWN_WITH_ZERO_TAX_WITH_ZERO_FEES,
  tradeIn: 2000,
};

export const WHEN_ZERO_DOWN_ZERO_TAX_ZERO_FEES = {
  MONTHLY_PAYMENT: 848.18,
  TOTAL_COST: 50891.07,
  TOTAL_INTEREST: 1891.07,
};

export const WHEN_DOWN_PAYMENT_ZERO_TAX_ZERO_FEES = {
  MONTHLY_PAYMENT: 813.56,
  TOTAL_COST: 50813.88,
  TOTAL_INTEREST: 1813.88,
};

export const WHEN_ZERO_DOWN_TAX_ZERO_FEES = {
  MONTHLY_PAYMENT: 901.2,
  TOTAL_COST: 54071.76,
  TOTAL_AMOUNT_TO_FINANCE: 52062.5,
  TOTAL_INTEREST: 2009.26,
};

export const WHEN_ZERO_DOWN_ZERO_TAX_AND_FEES = {
  MONTHLY_PAYMENT: 874.15,
  TOTAL_COST: 52448.96,
  TOTAL_AMOUNT_TO_FINANCE: 50500,
  TOTAL_INTEREST: 1948.96,
};

export const WHEN_DOWN_PAYMENT_TAX_AND_FEES = {
  MONTHLY_PAYMENT: 893.62,
  TOTAL_COST: 55617.38,
  TOTAL_AMOUNT_TO_FINANCE: 51625,
  TOTAL_INTEREST: 1992.38,
};


src/LeaseCalculator.js
import {
  TAX_ON_MONTHLY_PAYMENT,
  TAX_ON_SALES_PRICE,
  TAX_ON_TOTAL_LEASE_PAYMENT,
  MAKES,
  MANUFACTURER_FEES,
} from "./constants";

class LeaseCalculator {
  /*
    Validates required fields.
    Throws: Error | If there's an invalid required field
  */
  _validateData() {
    const requiredInputs = [
      "MSRP",
      "Selling Price",
      "Residual Value",
      "Money Factor",
    ];
    const requiredInput = [this.msrp, this.sellingPrice, this.rv, this.mf];
    requiredInput.forEach((field, i) => {
      if (!field || field === 0) {
        throw new Error(`Invalid Input: ${requiredInputs[i]}`);
      }
    });
  }

  /*
    Calculates the absolute and relative residual value of the vehicle
  */
  _calculateRV() {
    if (this.isRVPercent) {
      this.RVValue = this.msrp * (this.rv / 100);
      this.RVPercent = this.rv;
    } else {
      this.RVValue = this.rv;
      this.RVPercent = (this.rv / this.msrp) * 100;
    }
  }

  /*
    Converts the lease' money factor to APR.
    Note: To keep calculation accruate, we might need to truncate this value only on consumer read
    Returns: Number
  */
  _MFToAPR() {
    return this.mf * 2400;
  }

  /*
    Calculates the lease' monthly payment, APR, total cost, etc.

    make          Make of the vehicle, for calculating fees
    msrp          Required, MSRP of the vehicle
    sellingPrice  Required, negotiated price of the vehicle
    rv            Required, Residual value of the vehicle
    isRVPercent   Whether the rv is absolute value of a percentage of MSRP
    mf            Required, The money factor of the lease
    leaseTerm     The length of the lease in months.
    salesTax      The state's sales tax in percentage.
    totalFees     Total fees of the lease
    rebates       Total discount from dealer and manufacturer
    tradeIn       Total trade-in value
    downPayment   Down payment, if applicable
    taxMethod     Method of taxation to apply, based on state
  */
  calculate({
    make,
    msrp,
    sellingPrice,
    rv,
    isRVPercent = true,
    mf,
    leaseTerm = 36,
    salesTax = 0,
    totalFees = 0,
    rebates = 0,
    tradeIn = 0,
    downPayment = 0,
    taxMethod = TAX_ON_MONTHLY_PAYMENT,
  }) {
    this.make = make;
    this.msrp = msrp;
    this.sellingPrice = sellingPrice;
    this.rv = rv;
    this.isRVPercent = isRVPercent;
    this.mf = mf;
    this.leaseTerm = leaseTerm;
    this.salesTax = salesTax;
    this.totalFees = totalFees;
    this.rebates = rebates;
    this.tradeIn = tradeIn;
    this.downPayment = downPayment;
    this.taxMethod = taxMethod;

    this._validateData();

    this._calculateRV();
    const grossCapCost = this.sellingPrice + this.totalFees;
    const capCostReduction = this.downPayment + this.rebates + this.tradeIn;
    // Net Cap Cost. By default, acquisition fee is capitalized
    const adjustedCapCost =
      grossCapCost - capCostReduction + this.getAcquisitionFee();
    const depreciation = adjustedCapCost - this.RVValue;
    const basePayment = depreciation / this.leaseTerm;
    const rentCharge = (adjustedCapCost + this.RVValue) * this.mf;
    this.apr = this._MFToAPR();
    this.monthlyPaymentPreTax = basePayment + rentCharge;
    this.monthlyPayment = this.calculateMonthlyPaymentWithTax();
    this.driveOffPayment = this.getDriveOffPayment();
  }

  /*
    Gets the residual value of the lease in percentage 
    Returns: Number
  */
  getRVPercentage() {
    return Math.round(this.RVPercent);
  }

  /*
    Gets the residual value of the lease
    Returns: Number
  */
  getRVValue() {
    return Number.parseFloat(this.RVValue.toFixed(2));
  }

  /*
    Gets the monthly payment of the lease, not including taxes
    Returns: Number
  */
  getMonthlyPaymentPreTax() {
    return Math.round(this.monthlyPaymentPreTax * 100) / 100;
  }

  /*
    Gets the monthly payment of the lease, including taxes
    Returns: Number
  */
  getMonthlyPayment() {
    return Math.round(this.monthlyPayment * 100) / 100;
  }

  /*
    Gets the discount off of the MSRP, in percentage
    Returns: Number
  */
  getDiscountOffMsrpPercentage() {
    const offMsrp = this.msrp - this.sellingPrice;
    const offMsrpPercentage = (offMsrp / this.msrp) * 100;
    return Math.round(offMsrpPercentage * 100) / 100;
  }

  /*
    Gets the percentage of the monthly payment out of the MSRP
    Returns: Number
  */
  getMonthlyPaymentToMsrpPercentage() {
    const msrpPercentage = (this.monthlyPayment / this.msrp) * 100;
    return Math.round(msrpPercentage * 100) / 100;
  }

  /*
    Gets the total cost of the lease
    Returns: Number
  */
  getTotalLeaseCost() {
    const totalCost =
      // First monthly payment is included in drive-off
      this.monthlyPayment * (this.leaseTerm - 1) +
      this.getDriveOffPayment() +
      this.getDispositionFee();
    return Math.round(totalCost * 100) / 100;
  }

  /*
    Gets the APR value of the lease
    Returns: Number
  */
  getAPR() {
    return Math.round(this.apr * 100) / 100;
  }

  /*
    Gets the acquisition fee value by brand. If no brand sepcified, returns 0;
    Returns: Number
  */
  getAcquisitionFee() {
    const make = MAKES.filter((m) => m.displayName === this.make);
    if (make.length === 0) {
      return 0;
    }

    return MANUFACTURER_FEES.filter(
      (manufacturer) => manufacturer.makeId === make[0].id
    )[0].acquisitionFee;
  }

  /*
    Gets the disposition fee value by brand. If no brand sepcified, returns 0;
    Returns: Number
  */
  getDispositionFee() {
    const make = MAKES.filter((m) => m.displayName === this.make);
    if (make.length === 0) {
      return 0;
    }

    return MANUFACTURER_FEES.filter(
      (manufacturer) => manufacturer.makeId === make[0].id
    )[0].dispositionFee;
  }

  /*
    Calculates monthly payment based on method of taxation
    Returns: Number
  */
  calculateMonthlyPaymentWithTax() {
    if (this.taxMethod === TAX_ON_MONTHLY_PAYMENT)
      return this.monthlyPaymentPreTax * (1 + this.salesTax / 100);
    return this.monthlyPaymentPreTax;
  }

  /*
    Calculates the drive-off tax based on method of taxation
    Returns: Number
  */
  calculateDriveOffTaxes() {
    let taxableAmount;
    if (this.taxMethod === TAX_ON_MONTHLY_PAYMENT) {
      taxableAmount = this.downPayment + this.totalFees;
    } else if (this.taxMethod === TAX_ON_SALES_PRICE) {
      taxableAmount = this.sellingPrice;
    } else if (this.taxMethod === TAX_ON_TOTAL_LEASE_PAYMENT) {
      taxableAmount =
        this.monthlyPaymentPreTax +
        this.downPayment +
        this.totalFees +
        this.getAcquisitionFee();
    }

    // return taxableAmount * (1 + this.salesTax / 100);
    return taxableAmount * (this.salesTax / 100);
  }

  /*
    Calculates total drive-off payment
    Returns: Number
  */
  getDriveOffPayment() {
    const driveOffTaxes = this.calculateDriveOffTaxes();
    return (
      driveOffTaxes +
      this.downPayment +
      this.totalFees +
      // First month payment
      this.monthlyPayment
    );
  }
}

export default LeaseCalculator;


src/constants.ts
export enum TaxationMethod {
  TAX_ON_MONTHLY_PAYMENT = 1,
  TAX_ON_SALES_PRICE,
  TAX_ON_TOTAL_LEASE_PAYMENT,
}

export const MAKES = [
  { id: 1, displayName: "Acura", acquisitionFee: 595 },
  {
    id: 2,
    displayName: "Alfa Romeo",
    acquisitionFee: 595,
    dispositionFee: 395,
  },
  { id: 3, displayName: "Audi", acquisitionFee: 895, dispositionFee: 495 },
  {
    id: 4,
    displayName: "Bmw",
    acquisitionFee: 925,
    dispositionFee: 495,
    mfReductionUnit: 0.00007,
  },
  { id: 5, displayName: "Buick", acquisitionFee: 695, dispositionFee: 495 },
  { id: 6, displayName: "Cadillac", acquisitionFee: 695, dispositionFee: 595 },
  { id: 7, displayName: "Chevrolet", acquisitionFee: 695, dispositionFee: 395 },
  { id: 8, displayName: "Chrysler", acquisitionFee: 595, dispositionFee: 395 },
  { id: 9, displayName: "Dodge", acquisitionFee: 595, dispositionFee: 395 },
  { id: 10, displayName: "Fiat", acquisitionFee: 595, dispositionFee: 395 },
  { id: 11, displayName: "Ford", acquisitionFee: 645, dispositionFee: 395 },
  { id: 12, displayName: "Genesis", acquisitionFee: 750, dispositionFee: 400 },
  { id: 13, displayName: "GMC", acquisitionFee: 695, dispositionFee: 495 },
  { id: 14, displayName: "Honda", acquisitionFee: 595, dispositionFee: 0 },
  { id: 15, displayName: "Hyundai", acquisitionFee: 650, dispositionFee: 400 },
  {
    id: 16,
    displayName: "Infiniti",
    acquisitionFee: 795,
    dispositionFee: 395,
    mfReductionUnit: 0.0001,
  },
  { id: 17, displayName: "Jaguar", acquisitionFee: 895, dispositionFee: 300 },
  { id: 18, displayName: "Jeep", acquisitionFee: 595, dispositionFee: 395 },
  { id: 19, displayName: "Kia", acquisitionFee: 650, dispositionFee: 400 },
  {
    id: 20,
    displayName: "land rover",
    acquisitionFee: 895,
    dispositionFee: 300,
  },
  {
    id: 21,
    displayName: "Lexus",
    acquisitionFee: 795,
    dispositionFee: 350,
    mfReductionUnit: 0.00008,
  },
  { id: 22, displayName: "Lincoln", acquisitionFee: 645, dispositionFee: 395 },
  { id: 23, displayName: "Mini", acquisitionFee: 925, dispositionFee: 350 },
  { id: 24, displayName: "Mazda", acquisitionFee: 650, dispositionFee: 350 },
  {
    id: 25,
    displayName: "Mercedes benz",
    acquisitionFee: 795,
    dispositionFee: 595,
    mfReductionUnit: 0.00007,
  },
  { id: 26, displayName: "Mitsubishi", acquisitionFee: 595, dispositionFee: 0 },
  { id: 27, displayName: "Nissan", acquisitionFee: 696, dispositionFee: 395 },
  { id: 28, displayName: "Ram", acquisitionFee: 595, dispositionFee: 395 },
  { id: 31, displayName: "Subaru", acquisitionFee: 595, dispositionFee: 300 },
  { id: 32, displayName: "Tesla", acquisitionFee: 695, dispositionFee: 395 },
  {
    id: 33,
    displayName: "Toyota",
    acquisitionFee: 650,
    dispositionFee: 350,
    mfReductionUnit: 0.00008,
  },
  {
    id: 34,
    displayName: "Volkswagen",
    acquisitionFee: 699,
    dispositionFee: 395,
  },
  { id: 35, displayName: "Volvo", acquisitionFee: 995, dispositionFee: 350 },
];
